Add Static Equilibrium Tutorial

h2o-DS opened the pull request:
This replaces #56 

I wrote this tutorial for introducing vector algebra for solving static equilibrium physics problems. Uses NumPy for linear algebra and matplotlib for visualizations.

melissawm commented:
Thank you, @h2o-DS ! I'll start reviewing soon :+1:

melissawm reviewed (changes requested):
Hi @h2o-DS ! I haven't finished my first review (only made it as far as I've commented, but will continue tomorrow). I've made a number of comments, some of them are pretty minor. I do see one big issue: the image used in the "Finding values with physical properties" section. It is very large and to me it's not very clear what's going on. We can work together to find a better solution if you want. 

Also, if you have any questions or don't understand my comments, please let me know. 

Thanks!

cooperrc commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,392 @@
+
+Use matplotlib to visualize these forces.
+
+```{code-cell} ipython3
+fig = plt.figure()
+
+d3 = fig.gca(projection='3d')
+
+d3.set_xlim(-1,1)
+d3.set_ylim(-1,1)
+d3.set_zlim(-1,1)
+
+x, y, z = np.array([0,0,0]) # defining the point of application.  Make it the origin
+
+u, v, w = forceA # breaking the force vector into individual components
+d3.quiver(x,y,z,u,v,w) # plotting forceA
+
+u, v, w = forceB
+d3.quiver(x,y,z,u,v,w) # plotting forceB
+
+plt.show()
+```
````
thanks @melissawm, here are links for `quiver`:
- https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.quiver.html
- https://matplotlib.org/3.3.4/gallery/mplot3d/quiver3d.html

cooperrc commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,392 @@
+
+In order to simplify this problem, you can add them together to find the sum of forces.
+
+```{code-cell} ipython3
+forceC = forceA + forceB
+print('Force C =', forceC)
+```
````
Great idea!
In Python, you would have 
```
forceC = []
for i in range(len(forceA)):
    forceC[i] = forceA[i] +forceB[i]
```

cooperrc commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,392 @@
+print('Force C =', forceC)
+```
+
+Force C now acts as a single force that represents both A and B.
+You can plot it to see the result.
````
> What does this result mean (for the reader)? Can it be compared to A and B? Is it possible to do a plot with all three vectors, maybe using different colors?

That would look nice. 
e.g. 
```
x, y, z = np.array([0,0,0])
u, v, w = forceA
d3.quiver(x,y,z,u,v,w)
x, y, z = forceA
u, v, w = forceB
d3.quiver(x,y,z,u,v,w)
x, y, z = np.array([0,0,0])
u, v, w = forceC
d3.quiver(x,y,z,u,v,w)
```
There are better ways to show my example above though, e.g. 
```
fig = plt.figure()
d3 = fig.gca(projection='3d')
d3.set_xlim(-4,4)
d3.set_ylim(-4,4)
d3.set_zlim(-4,4)
x, y, z = np.vstack([np.zeros(3), a, np.zeros(3)]).T
u, v, w = np.vstack([a, b, c]).T

d3.quiver(x, y, z, u, v, w, color=['r', 'b', 'g'])
```
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,384 @@
+
+d3 = fig.gca(projection='3d')
+
+d3.set_xlim(-1,1)
+d3.set_ylim(-1,1)
+d3.set_zlim(-1,1)
+
+x, y, z = np.array([0,0,0])
+
+u, v, w = forceC
+d3.quiver(x,y,z,u,v,w)
+
+plt.show()
+```
+
+There is now a single vector at the origin.
````
```suggestion
d3.set_xlim(-1,1)
d3.set_ylim(-1,1)
d3.set_zlim(-1,1)

x, y, z = np.array([0,0,0])

u, v, w = forceA
d3.quiver(x,y,z,u,v,w,color='r')
u, v, w = forceB
d3.quiver(x,y,z,u,v,w,color='b')
u, v, w = forceC
d3.quiver(x,y,z,u,v,w,color='g')

plt.show()
```

There is now a larger green vector that is a combination of both force A and force B.
This represents both of them acting at once.
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,384 @@
+x, y, z = np.array([0,0,0]) # defining the point of application.  Make it the origin
+
+u, v, w = forceA # breaking the force vector into individual components
+d3.quiver(x,y,z,u,v,w) # plotting forceA
+
+u, v, w = forceB
+d3.quiver(x,y,z,u,v,w) # plotting forceB
```
```suggestion
d3.quiver(x,y,z,u,v,w,color='r') # plotting forceA

u, v, w = forceB
d3.quiver(x,y,z,u,v,w,color='b') # plotting forceB

```

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,384 @@
+
+solving for R_x, R_y, and R_z gives you a vector $R$ of $(-1, -1, 0)$.
+
+Plot this and see if it works.
```
```suggestion
If plotted, the forces seen in prior examples should be nullified.
Only if there is no force remaining is the system considered to be in equilibrium.
```

h2o-DS commented:
@melissawm Thank you for your criticisms! I believe I have addressed the suggestions and comments you have made. The review can continue whenever you have the time.

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+0+0+R_z=0
+\end{cases}$$
+
+solving for R_x, R_y, and R_z gives you a vector $R$ of $(-1, -1, 0)$.
```
```suggestion
solving for $R_x$, $R_y$, and $R_z$ gives you a vector $R$ of $(-1, -1, 0)$.

```

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+A simple example of this would be from a force applied to a stationary pole secured in the ground.
+The pole does not move, so it must apply a reaction force.
+The pole also does not rotate, so it must also be creating a reaction moment.
+Solve for both.
```
Can you expand a bit on what "Solve for both" means here?

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+Lets say a 5N force is applied 2m above the base of the pole.
+
+```{code-cell} ipython3
+f = 5
+L = 2
+
+R = 0 - f
+M = 0 - f*L
+print('Reaction force =',R)
+print('Reaction moment =',M)
````
```suggestion
f = 5  # Force in newtons
L = 2  # Length of the pole

R = 0 - f
M = 0 - f*L
print('Reaction force =',R)
print('Reaction moment =',M)

```

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+The pole also does not rotate, so it must also be creating a reaction moment.
+Solve for both.
+
+Lets say a 5N force is applied 2m above the base of the pole.
```
Is this force being applied perpendicularly? (that's what I understand from the next example, but it would be nice to make it clearer here)

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+Let's say that instead of a force acting perpendicularly to the beam, a force was applied to our pole through a wire that was also attached to the ground.
+Given the tension in this cord, all you need to solve this problem are the physical locations of these objects.
+
+![image.png](tutorial-static_equilibrium/static_eqbm-fig01.png)
```
Much better!

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+Denote the base of the pole as the origin.
+Now, say the cord is attached to the ground 3m in the x direction and attached to the pole 2m up, in the z direction.
+
+Define these points in space as numpy arrays, and then use those arrays to find directional vectors.
```
```suggestion
Define these points in space as NumPy arrays, and then use those arrays to find directional vectors.

```

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+Define these points in space as numpy arrays, and then use those arrays to find directional vectors.
+
+```{code-cell} ipython3
+poleBase = np.array([0,0,0])
+cordBase = np.array([3,0,0])
+cordConnection = np.array([0,0,2])
+
+poleDirection = cordConnection - poleBase
+print('Pole direction =',poleDirection)
+cordDirection = cordBase - cordConnection
+print('Cord direction =',cordDirection)
````
PEP8:
```suggestion
poleBase = np.array([0, 0, 0])
cordBase = np.array([3, 0, 0])
cordConnection = np.array([0, 0, 2])

poleDirection = cordConnection - poleBase
print('Pole direction =', poleDirection)
cordDirection = cordBase - cordConnection
print('Cord direction =', cordDirection)

```

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+
+```{code-cell} ipython3
+cordUnit = cordDirection/np.linalg.norm(cordDirection)
+print('Cord unit vector =',cordUnit)
````
```suggestion
print('Cord unit vector =', cordUnit)

```

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+
+You can then multiply this direction with the magnitude of the force in order to find the force vector.
+
+Let's say the cord has a tension of 5N
```
```suggestion
Let's say the cord has a tension of 5N:

```

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+```{code-cell} ipython3
+cordTension = 5
+forceCord = cordUnit * cordTension
+print('Force from the cord =',forceCord)
````
```suggestion
print('Force from the cord =', forceCord)

```

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+
+```{code-cell} ipython3
+momentCord = np.cross(forceCord, poleDirection)
+print('Moment from the cord =',momentCord)
````
```suggestion
print('Moment from the cord =', momentCord)

```

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+Now all you need to do is find the reaction force and moment.
+
+```{code-cell} ipython3
+equilibrium = np.array([0,0,0])
+R = equilibrium - forceCord
+M = equilibrium - momentCord
+print("Reaction force =",R)
+print("Reaction moment =",M)
````
```suggestion
equilibrium = np.array([0, 0, 0])
R = equilibrium - forceCord
M = equilibrium - momentCord
print("Reaction force =", R)
print("Reaction moment =", M)

```

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+```
+
+## Another Example
+Let's look at a slightly more complicated model.  In this example you will be observing a beam with two cables and an applied force.  This time you need to find both the tension in the cords and the reaction forces of the beam.
+
+[Vector Mechanics for Engineers: Statics](https://www.mheducation.com/highered/product/vector-mechanics-engineers-statics-beer-johnston/M9780077687304.html)
+
+Problem 4.106
````
```suggestion
Let's look at a slightly more complicated model.  In this example you will be observing a beam with two cables and an applied force.  This time you need to find both the tension in the cords and the reaction forces of the beam. *(Source: [Vector Mechanics for Engineers: Statics](https://www.mheducation.com/highered/product/vector-mechanics-engineers-statics-beer-johnston/M9780077687304.html), Problem 4.106)*

```

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+
+![image.png](tutorial-static_equilibrium/problem4.png)
+
+Define distance a as 3m
```
I'm not sure if this is supposed to be here

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+As before, start by defining the location of each relevant point as an array.
+
+```{code-cell} ipython3
+A = np.array([0,0,0])
+B = np.array([0,3,0])
+C = np.array([0,6,0])
+D = np.array([1.5,0,-3])
+E = np.array([1.5,0,3])
+F = np.array([-3,0,2])
````
```suggestion
A = np.array([0, 0, 0])
B = np.array([0, 3, 0])
C = np.array([0, 6, 0])
D = np.array([1.5, 0, -3])
E = np.array([1.5, 0, 3])
F = np.array([-3, 0, 2])

```

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+UnitBE = BE/np.linalg.norm(BE)
+UnitCF = CF/np.linalg.norm(CF)
+
+RadBD = np.cross(AB,UnitBD)
+RadBE = np.cross(AB,UnitBE)
+RadCF = np.cross(AC,UnitCF)
```
```suggestion
RadBD = np.cross(AB, UnitBD)
RadBE = np.cross(AB, UnitBE)
RadCF = np.cross(AC, UnitCF)

```

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+RadCF = np.cross(AC,UnitCF)
+```
+
+This lets you represent the forces acting on the system as
````
I'm not sure I understand this notation - what is `$T_{BD}$` and `$T_{BE}$`?

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+\end{array}
+\right]$
+
+Where T is the tesnion in the respective cord and R is the reaction force in a respective direction. Then you just have six equations:
```
```suggestion
Where $T$ is the tension in the respective cord and $R$ is the reaction force in a respective direction. Then you just have six equations:

```

melissawm commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,389 @@
+
+Where T is the tesnion in the respective cord and R is the reaction force in a respective direction. Then you just have six equations:
+
+$\sum F_{x} = 0 = T_{BE}/3+T_{BD}/3-195+R_{x}$
+
+$\sum F_{y} = 0 = T_{BE}*-2/3-T_{BD}*2/3-390+R_{y}$
+
+$\sum F_{z} = 0 = T_{BE}*-2/3+T_{BD}*2/3+130+R_{z}$
+
+$\sum M_{x} = 0 = 780+2*T_{BE}-2*T_{BD}$
+
+$\sum M_{z} = 0 = 1170-T_{BE}-T_{BD}$
```
```suggestion
$\sum F_{x} = 0 = T_{BE}/3+T_{BD}/3-195+R_{x}$

$\sum F_{y} = 0 = (-\frac{2}{3})T_{BE}-\frac{2}{3}T_{BD}-390+R_{y}$

$\sum F_{z} = 0 = (-\frac{2}{3})T_{BE}+\frac{2}{3}T_{BD}+130+R_{z}$

$\sum M_{x} = 0 = 780+2T_{BE}-2T_{BD}$

$\sum M_{z} = 0 = 1170-T_{BE}-T_{BD}$

```

melissawm commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,389 @@
+1. [Vector Mechanics for Engineers: Statics (Beer & Johnston & Mazurek)](https://www.mheducation.com/highered/product/vector-mechanics-engineers-statics-beer-johnston/M9780077687304.html)
+2. [NumPy Reference](https://numpy.org/doc/stable/reference/)
+
+```{code-cell} ipython3
+
+```
````
```suggestion
```

melissawm reviewed (changes requested):
Hello, @h2o-DS ! Thanks for that - I added a few more suggestions and would like to note that some of the code cells are mixed with text - can you separate those so the final formatting/rendering in html is correct? Cheers!

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,393 @@
+
+![image.png](tutorial-static_equilibrium/problem4.png)
+
+Define distance a as 3m
```
```suggestion
Define distance *a* as 3 meters

```

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,394 @@
+
+plt.legend()
+plt.show()
+However, the goal is equilibrium.
```
```suggestion
However, the goal is equilibrium.
```
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,397 @@
+A simple example of this would be from a force applied to a stationary pole secured in the ground.
+The pole does not move, so it must apply a reaction force.
+The pole also does not rotate, so it must also be creating a reaction moment.
+Solve for both.
```
```suggestion
Solve for both the reaction force and moments.
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,397 @@
+The pole also does not rotate, so it must also be creating a reaction moment.
+Solve for both.
+
+Lets say a 5N force is applied 2m above the base of the pole.
```
```suggestion
Lets say a 5N force is applied perpendicularly 2m above the base of the pole.
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,392 @@
+A simple example of this would be from a force applied to a stationary pole secured in the ground.
+The pole does not move, so it must apply a reaction force.
+The pole also does not rotate, so it must also be creating a reaction moment.
+Solve for both.
```
```suggestion
Solve for both of these forces.
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,392 @@
+The pole also does not rotate, so it must also be creating a reaction moment.
+Solve for both.
+
+Lets say a 5N force is applied 2m above the base of the pole.
```
```suggestion
Lets say a 5N force is applied perpendicularly 2m above the base of the pole.
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
```diff
@@ -0,0 +1,392 @@
+
+![image.png](tutorial-static_equilibrium/problem4.png)
+
+Define distance a as 3m
```
```suggestion
Define distance *a* as 3m
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,392 @@
+RadCF = np.cross(AC,UnitCF)
+```
+
+This lets you represent the forces acting on the system as
````
```suggestion
This lets you represent the tension (T) and reaction (R) forces acting on the system as
```

h2o-DS commented on content/tutorial-static_equilibrium.md:
````diff
@@ -0,0 +1,397 @@
+
+```
+
+This lets you represent the tension and reaction forces acting on the system as
````
```suggestion
This lets you represent the tension (T) and reaction (R) forces acting on the system as
```

h2o-DS commented:
@melissawm I believe I have finally gotten around to addressing all of your comments and suggestions. If there is anything else let me know. Thank you for all your help

melissawm commented:
Hello, @h2o-DS ! I pushed some formatting fixes just to get the site to build correctly but we can put this in, and if there's anything else we want to add we can do it on a follow-up. Thank you for your contribution and feel free to do it again :) Cheers!
